### ISSUE TITLE
Cloud-init cannot find DataSourcenocloud module

### ISSUE THREAD

**Issue by Stalis** [canonical/cloud-init#4302]
# Bug report
Ubuntu 22.04 cannot installed through cloud-init. In logs i see next line:
`2023-07-28 06:30:41,833 ERROR cloudinit.sources:1010 Could not import DataSourcenocloud. Does the DataSource exist and is it importable?`
It doesn't work with ds=nocloud-net too - same error, but for `DataSourcenocloud-net` module
But, strange thing, i see data from `user-data` in `instance-data-sensitive.json` file

## Steps to reproduce the problem
1. Run QEMU VM from Ubuntu 22.04 ISO [Image Download link](https://cdimage.ubuntu.com/ubuntu-server/jammy/daily-live/current/jammy-live-server-amd64.iso) and iso with user-data file(link below) and empty meta-data
[user-data.txt](https://github.com/canonical/cloud-init/files/12190992/user-data.txt)
```
# Command example:
/usr/bin/qemu-system-x86_64 \
  -cpuhost -name ubuntu-22.04-amd64.qemu.stable.base.qcow2 \
  -machine type=pc,accel=kvm \
  -netdev user,id=user.0,hostfwd=tcp::3954-:22 \
  -device virtio-scsi-pci,id=scsi0 \
  -device scsi-hd,bus=scsi0.0,drive=drive0 \
  -device virtio-net,netdev=user.0 \
  -m 2048 -smp cpus=2,sockets=1,cores=2,threads=1 \
  -vnc 127.0.0.1:17 \
  -drive if=none,file=image.qcow2,id=drive0,cache=writeback,discard=ignore,format=qcow2 \
  -drive file=Ubuntu-22.04.iso,media=cdrom \
  -drive file=cloud-init.iso,media=cdrom \
  -boot once=d
```
2. Run system with command
```
linux /casper/vmlinuz --- autoinstall ds="nocloud"
initrd /casper/initrd
boot
```
3. And... Nothing happens

## Environment details
- Cloud-init version: cloud-init 23.1.1-0ubuntu0~22.04.1
- Operating System Distribution: Ubuntu Server 22.04.1
- Cloud provider, platform or installer type: QEMU through Packer on amd64 system

## cloud-init logs
Please pay attention on that line: cloud-init-logs-2023-07-28/var/log/installer/subiquity-server-debug.log:59
[cloud-init.tar.gz](https://github.com/canonical/cloud-init/files/12190950/cloud-init.tar.gz)

**Comment by Stalis** [comment]
Tried to run same, but use `ds=nocloud`(w/o dquotes) - received same behaviour

**Comment by blackboxsw** [comment]
Thank you @Stalis for the bug and making cloud-init better. Unfortunately, cloud-init collect-logs doesn't collect your /var/log/installer/subiquity-server-debug.log so it's not attached. I presume though that because there are no tracebacks in /var/log/cloud-init.log that this is an Ubuntu Live installer (subiquity) bug related to https://github.com/canonical/cloud-init/issues/4158 which is not a cloud-init bug, but an Ubuntu live installer bug. I expect on your system you would see a Traceback in /var/log/installler/subiquity-server-debug.log. Note this is is a known problem due to a recent SRU of cloud-init 23.2.1 to Jammy that was just fixed last week in the ubuntu live installer (subiquity) https://bugs.launchpad.net/subiquity/+bug/2022102/comments/5 as well. If this is not a symptom of jammy daily that isn't resolved in a couple of days. Feel free to open a new subiquity bug or re-open this cloud-init bug.

**Comment by dbungert** [comment]
A fix for this issue has been included in the daily builds of what will be 22.04.3.  Please use build 20230801.1 or newer.

**Comment by DLAtWork** [comment]
Sorry to comment on a closed bug but I've encountered what appears to be this issue testing automated installs using the Ubuntu 22.04.3 LTS Live Server Installer ISO that was released on August 10th.

Package version: cloud-init 23.2.1-0ubuntu0~22.04.1

Sanitized error from the installer logs:
```sh
Cloud-init v. 23.2.1-0ubuntu0~22.04.1 running 'init-local' at Fri, 11 Aug 2023 21:54:06 +0000. Up 22.09 seconds.
2023-08-11 21:54:06,085 - main.py[WARNING]: contents of '/dev/null' did not start with b'#cloud-config'
2023-08-11 21:54:06,096 - util.py[WARNING]: failed stage init-local
failed run of stage init-local
------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 766, in status_wrapper
    ret = functor(name, args)
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 384, in main_init
    init.fetch(existing=existing)
  File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 436, in fetch
    return self._get_data_source(existing=existing)
  File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 327, in _get_data_source
    (ds, dsname) = sources.find_source(
  File "/usr/lib/python3/dist-packages/cloudinit/sources/__init__.py", line 996, in find_source
    ds_list = list_sources(cfg_list, ds_deps, pkg_list)
  File "/usr/lib/python3/dist-packages/cloudinit/sources/__init__.py", line 1043, in list_sources
    m_locs, _looked_locs = importer.find_module(
  File "/usr/lib/python3/dist-packages/cloudinit/importer.py", line 70, in find_module
    if not importlib.util.find_spec(full_path):
  File "/usr/lib/python3.10/importlib/util.py", line 94, in find_spec
    parent = __import__(parent_name, fromlist=['__path__'])
ModuleNotFoundError: No module named 'DataSourcenocloud-net;s=https://LOCAL.DEPLOY.SERVER'
------------------------------------------------------------
```

Sanitized installer /proc/cmd:
```sh
initrd=initrd noprompt hostname=LOCAL.HOST.NAME BOOTIF=01-XX:XX:XX:XX:XX ip=dhcp netboot=url url=https://LOCAL.DEPLOY.SERVER/ubuntu-releases/22.04.3/ubuntu-22.04.3-live-server-amd64.iso BOOT=casper autoinstall cloud-config-url=/dev/null ds=nocloud-net;s=https://XXXX/SERVER-DEPLOY-FILES/;h=LOCAL.HOST.NAME; nopersistent fsck.mode=skip --- 
```

We PXE boot the system and it downloads the ISO to RAM and then kicks off the automated install based on the kernel cmdline parameters. In this case the installer boots but fails to start the automated installation due to this error.

**Comment by DLAtWork** [comment]
OK, after some more searching I see that the updated package with the fix for #4271 is 23.2.2-0ubuntu0~22.04.1, per the package change log:
http://changelogs.ubuntu.com/changelogs/pool/main/c/cloud-init/cloud-init_23.2.2-0ubuntu0~22.04.1/changelog
https://raw.githubusercontent.com/canonical/cloud-init/23.2.2/ChangeLog

It looks like the Jammy Live Server Daily ISO from 08/12 includes cloud-init 23.2.2:
https://cdimage.ubuntu.com/ubuntu-server/jammy/daily-live/20230812/jammy-live-server-amd64.manifest

So this isn't an issue with a regression, it's an issue of an older package, 23.2.1-0ubuntu0~22.04.1, being shipped on the 22.04.3 LTS release ISO:
https://releases.ubuntu.com/jammy/ubuntu-22.04.3-live-server-amd64.manifest

Odds on the release ISO getting a respin to fix this issue?

**Comment by dbungert** [comment]
> ```shell
> ds=nocloud-net;s=https://XXXX/SERVER-DEPLOY-FILES/;h=LOCAL.HOST.NAME
> ```

I do not believe that Subiquity does anything meaningful with the local-hostname part of this nocloud declaration.  Please retry with `ds=nocloud-net;s=https://XXXX/SERVER-DEPLOY-FILES/`

To set the hostname you would need a
```yaml
autoinstall:
  ...
  identity:
    ...
    hostname: LOCAL.HOST.NAME
```

**Comment by DLAtWork** [comment]
@dbungert Thanks for checking a comment on a closed issue, much appreciated.

Our deployment system is home grown (a quick temporary solution based on d-i, that is not so temporary, and got updated to work with subiquity). It doesn't have a means to modify `user-data` at deployment time. We inject the system hostname on the kernel command line.

I did a quick test removing the `;h=LOCAL.HOST.NAME` option from `ds=` , just to see if that would get the install going again and we could deal with any hostname later. Same result as before, unfortunately, with the original 08/10 release ISO.

I substituted the release 22.04.3 live server ISO for the 08/14 daily build with `cloud-init 23.2.2-0ubuntu0~22.04.1` and the install works as expected with our original boot line.

We've got an automatically sync'ed local mirror of https://releases.ubuntu.com/ we had been using with Subiquity based installers, but I guess we can roll with a daily ISO build for now since it resolves the issue.

Again, thanks for taking the time to respond to a closed issue.

**Comment by DLAtWork** [comment]
@dbungert, sorry about tagging you here and on the mailing list.

TL;DR: Removing the `;` at the end of `s=URL` fixed the parser issue but broke the deployed hostname. We'll stick to the 08/14 daily ISO as it works at this point.

Textwall:
Just to follow up on this, I went back and retested. I realized that I left the `;` at the end of the `s=URL` value.   
Removed that `;` and the installation now makes it past parsing the DataSource and completes the install.  
However, at the end of the installation the deployed hostname is now `CIMISSINGJINJAVARlocalhostname`. Ha! :rofl:   

Did some more digging and it does look like the `h=LOCAL.HOST.NAME` value ends up in an `/autoinstall.yml` file after `cloud-config` parses `user-data` as a Jinja template. In the `user-data` `autoinstall:identity` key, the `hostname` value is set to `{{ ds.meta_data.local_hostname }}`. So we are kind updating `user-data` during install but the values are still injected on the command line.

But this is all moot anyway since the issue is fixed in `cloud-init 23.2.2`. 

We'll just keep using the 08/14 daily ISO at this point, since it behaves as expected with our existing workflow.

**Comment by Silvenga** [comment]
@DLAtWork just want to thank you for your verboseness, it helped quite a bit - also ran into this, pulling my hair out why 22.04.3 was refusing to install with known good configuration.

(also discovered that Subiquity crashes when the hardware clock is bad, but that's a separate issue)

**Comment by DLAtWork** [comment]
@Silvenga Glad it helped. We've definitely had challenges moving from the legacy Debian Installer based deployment to Casper/Cloud-Init/Subiquity/Cloud-Init. It most certainly isn't geared for non-image based deployments to VMs and bare metal but, unfortunately, that's our use case so we're making it work as best we can.

We'll be looking at using Debian for our deployments in the mid-future given all the challenges with the increasing size of the deployment ISO and Subiquity being a moving target (even during the installation! Thanks `snapd`!) the last few years.

**Comment by Silvenga** [comment]
TBH, same on all accounts - we are also geared to non-image based deployments. We've also been considering moving to Debian for the exact same reason (although, Flatcar is currently a maybe, although, that's been a moving target too). At this point, since a lot of our VM's are on 18.04 - it might happen sooner than later.

We skipped Ubuntu 20.04 due to the move to Subiquity - it took us so long to figure out how to reproduce our installs with Cloud-init/Autoinstall, coming from legacy Debian - but since deploying some VM's in 22.04 - we've had nothing but problems.

I see the beauty of Subiquity - but it's definitely a departure from installing Debian with a 100MB net-installer.
